Solidity 0.8.18 Release blogpost (#204)
ethereum/solidity-blog: _posts/2023-02-01-solidity-0.8.18-release-announcement.md

---
layout: post
published: true
title: 'Solidity 0.8.18 Release Announcement'
date: '2023-02-01'
author: Solidity Team
category: Releases
---

> [Solidity v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18) introduces support for the [Paris upgrade](https://blog.ethereum.org/2022/08/24/mainnet-merge-announcement) and [user-defined operators](https://forum.soliditylang.org/t/feature-preview-user-defined-operators/1435)!

Author: ```suggestion
Introducing the Newest Version of the Solidity Compiler!
We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18). 
This latest version includes a range of improvements and it also introduces support for the [Paris upgrade](https://blog.ethereum.org/2022/08/24/mainnet-merge-announcement)!
```

### Compiler Features

* No CBOR metadata
    - ``` example code ```
* The support to the Paris hard fork includes: 
* Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
* Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.

*Note*: `block.difficulty` is planned to be completely removed in Solidity version `0.9.0`.

## Full Changelog

#### Important fixes
*
*

### Bugfixes
* Docs build crash with sphinx-a4doc >= 1.4.0 #13820
* Required boost version in the docs does not match the code after recent changes #13870
* AST source locations for named mapping keys. #13864
* 

We especially thank all the contributors that made this release possible:
LIST

Download the new version of Solidity here: [v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18).

> If you want to perform a source build, please only use `solidity_0.8.18.tar.gz` and not the zip provided by github directly.

Author: ```suggestion
If you want to perform a source build, please only use `solidity_0.8.18.tar.gz` and not the zip provided by github directly.

Get ready to unleash your creativity and build even more impactful decentralized applications with the latest version of the Solidity Compiler. Happy coding!
```

### Compiler Features

* No CBOR metadata
    - ``` example code ```
* The support to the Paris hard fork includes: 
* Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
* Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.

*Note*: `block.difficulty` is planned to be completely removed in Solidity version `0.9.0`.

## Full Changelog

#### Important fixes
*
*

### Bugfixes
* Docs build crash with sphinx-a4doc >= 1.4.0 #13820
* Required boost version in the docs does not match the code after recent changes #13870
* AST source locations for named mapping keys. #13864
* 

We especially thank all the contributors that made this release possible:
LIST

Download the new version of Solidity here: [v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18).
If you want to perform a source build, please only use `solidity_0.8.18.tar.gz` and not the zip provided by github directly.

> Get ready to unleash your creativity and build even more impactful decentralized applications with the latest version of the Solidity Compiler. Happy coding!

Author: ```suggestion
We encourage all Solidity developers to upgrade to version 0.8.18 to take advantage of these improvements and optimizations. Our team remains dedicated to continuously enhancing the Solidity platform to benefit the entire Ethereum ecosystem.

Get ready to unleash your creativity and build even more impactful decentralized applications with the latest version of the Solidity Compiler. Happy coding!
```

This latest version includes a range of improvements and it also introduces support for the [Paris upgrade](https://blog.ethereum.org/2022/08/24/mainnet-merge-announcement)!

### Compiler Features

* No CBOR metadata
    - ``` example code ```
* The support to the Paris hard fork includes: 
* Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
* Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.

*Note*: `block.difficulty` is planned to be completely removed in Solidity version `0.9.0`.

## Full Changelog

#### Important fixes
*
*

### Bugfixes
* Docs build crash with sphinx-a4doc >= 1.4.0 #13820
* Required boost version in the docs does not match the code after recent changes #13870
* AST source locations for named mapping keys. #13864
* 

We especially thank all the contributors that made this release possible:
LIST

> Download the new version of Solidity here: [v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18).
> If you want to perform a source build, please only use `solidity_0.8.18.tar.gz` and not the zip provided by github directly.

Author: ```suggestion
To upgrade to the latest version of the Solidity Compiler, simply follow the [installation instructions](https://docs.soliditylang.org/en/v0.8.18/installing-solidity.html) available in our documentation. Our team has made sure to provide detailed and straightforward steps to make the upgrade process as seamless as possible. If you have any questions or run into any issues during the upgrade process, don't hesitate to reach out to our [community Matrix channel](https://matrix.to/#/#ethereum_solidity:gitter.im). 
You can download the new version of Solidity here: [v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18).
If you want to perform a source build, please only use `solidity_0.8.18.tar.gz` and not the zip provided by GitHub directly.
```

---
layout: post
published: true
title: 'Solidity 0.8.18 Release Announcement'
date: '2023-02-01'
author: Solidity Team
category: Releases
---

Introducing the Newest Version of the Solidity Compiler!
We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18). 
This latest version includes a range of improvements and it also introduces support for the [Paris upgrade](https://blog.ethereum.org/2022/08/24/mainnet-merge-announcement)!

### Compiler Features

> * No CBOR metadata
>     - ``` example code ```

Reviewer: ````suggestion
* A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now it was only possible to exclude the hash of the JSON metadata file (`solc --metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.
````

> * A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now it was only possible to exclude the hash of the JSON metadata file (`solc --metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.

Reviewer 2: ```suggestion
* A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now, it was only possible to exclude the hash of the JSON metadata file (`solc --metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.
```

Reviewer: Oh, I also see that I used `solc --metadata-hash none`. Should have been just `--metadata-hash none`.

Author: ```suggestion
* A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now, it was only possible to exclude the hash of the JSON metadata file (` --metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.
```

Reviewer: Yes, that's what I meant.

Reviewer: Wait, you could also remove that extra space :)

Reviewer: ```suggestion
* A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now, it was only possible to exclude the hash of the JSON metadata file (`--metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.
```

> * The support to the Paris hard fork includes: 
> * Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
> * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.

Reviewer: ```suggestion
* The support to the Paris hard fork includes: 
    * Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
    * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.
```

> * Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
> * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.

Reviewer: ```suggestion
* Deprecation of global `block.difficulty` built-in in Solidity and removal of `difficulty()` instruction from inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
* Introduction of global `block.prevrandao` built-in in Solidity and `prevrandao()` instruction in inline assembly for EVM versions >= Paris.
```

Reviewer: This seems to have disappeared from the PR.

> * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.

Author: ```suggestion
* Deprecation of global `block.difficulty` built-in in Solidity and removal of `difficulty()` instruction from inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
* Introduction of global `block.prevrandao` built-in in Solidity and `prevrandao()` instruction in inline assembly for EVM versions >= Paris.
```

> *Note*: `block.difficulty` is planned to be completely removed in Solidity version `0.9.0`.

Author: ```suggestion
```

>     * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.

Reviewer: ```
* `selfdestruct` is now considered deprecated ([EIP-6049](https://eips.ethereum.org/EIPS/eip-6049)) and the compiler will warn about its use, both in Solidity, including inline assembly, and in Yul.
    There is currently no replacement, but its use is highly discouraged because it will eventually change its semantics and all contracts using it will be affected in some way.
```

> ## Full Changelog

Reviewer 3: This still missing. Should I add it?

Author: ```suggestion
## Full Changelog
Language Features:
 * Allow named parameters in mapping types.

Compiler Features:
 * Commandline Interface: Add ``--no-cbor-metadata`` that skips CBOR metadata from getting appended at the end of the bytecode.
 * Commandline Interface: Return exit code ``2`` on uncaught exceptions.
 * EVM: Deprecate ``block.difficulty`` and disallow ``difficulty()`` in inline assembly for EVM versions >= paris. The change is due to the renaming introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
 * EVM: Introduce ``block.prevrandao`` in Solidity and ``prevrandao()`` in inline assembly for EVM versions >= paris.
 * EVM: Set the default EVM version to "Paris".
 * EVM: Support for the EVM version "Paris".
 * Language Server: Add basic document hover support.
 * Natspec: Add event Natspec inheritance for devdoc.
 * Optimizer: Added optimization rule ``and(shl(X, Y), shl(X, Z)) => shl(X, and(Y, Z))``.
 * Parser: More detailed error messages about invalid version pragmas.
 * SMTChecker: Make ``z3`` the default solver for the BMC and CHC engines instead of all solvers.
 * SMTChecker: Support Eldarica as a Horn solver for the CHC engine when using the CLI option ``--model-checker-solvers eld``. The binary ``eld`` must be available in the system.
 * Solidity Upgrade Tool: Remove ``solidity-upgrade`` tool.
 * Standard JSON: Add a boolean field ``settings.metadata.appendCBOR`` that skips CBOR metadata from getting appended at the end of the bytecode.
 * TypeChecker: Warn when using deprecated builtin ``selfdestruct``.
 * Yul EVM Code Transform: Generate more optimal code for user-defined functions that always terminate a transaction. No return labels will be pushed for calls to functions that always terminate.
 * Yul Optimizer: Allow replacing the previously hard-coded cleanup sequence by specifying custom steps after a colon delimiter (``:``) in the sequence string.
 * Yul Optimizer: Eliminate ``keccak256`` calls if the value was already calculated by a previous call and can be reused.

Bugfixes:
 * Parser: Disallow several ``indexed`` attributes for the same event parameter.
 * Parser: Disallow usage of the ``indexed`` attribute for modifier parameters.
 * SMTChecker: Fix display error for negative integers that are one more than powers of two.
 * SMTChecker: Fix internal error on chain assignments using static fully specified state variables.
 * SMTChecker: Fix internal error on multiple wrong SMTChecker natspec entries.
 * SMTChecker: Fix internal error when a public library function is called internally.
 * SMTChecker: Fix internal error when deleting struct member of function type.
 * SMTChecker: Fix internal error when using user-defined types as mapping indices or struct members.
 * SMTChecker: Improved readability for large integers that are powers of two or almost powers of two in error messages.
 * TypeChecker: Fix bug where private library functions could be attached with ``using for`` outside of their declaration scope.
 * Yul Optimizer: Hash hex and decimal literals according to their value instead of their representation, improving the detection of equivalent functions.
```

## Full Changelog

#### Important fixes
*
*

### Bugfixes
* Docs build crash with sphinx-a4doc >= 1.4.0 #13820
* Required boost version in the docs does not match the code after recent changes #13870
* AST source locations for named mapping keys. #13864
* 

We especially thank all the contributors that made this release possible:
LIST

> To upgrade to the latest version of the Solidity Compiler, simply follow the [installation instructions](https://docs.soliditylang.org/en/v0.8.18/installing-solidity.html) available in our documentation. Our team has made sure to provide detailed and straightforward steps to make the upgrade process as seamless as possible. If you have any questions or run into any issues during the upgrade process, don't hesitate to reach out to our [community Matrix channel](https://matrix.to/#/#ethereum_solidity:gitter.im). 

Author: ```suggestion
To upgrade to the latest version of the Solidity Compiler, simply follow the [installation instructions](https://docs.soliditylang.org/en/v0.8.18/installing-solidity.html) available in our documentation. 
Our team has made sure to provide detailed and straightforward steps to make the upgrade process as seamless as possible. If you have any questions or run into any issues during the upgrade process, don't hesitate to reach out to our [community Matrix channel](https://matrix.to/#/#ethereum_solidity:gitter.im). 
```

You can download the new version of Solidity here: [v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18).

> If you want to perform a source build, please only use `solidity_0.8.18.tar.gz` and not the zip provided by GitHub directly.

Reviewer: ```suggestion
If you want to perform a source build, please only use [`solidity_0.8.18.tar.gz`](https://github.com/ethereum/solidity/releases/download/v0.8.18/solidity_0.8.18.tar.gz) and not the zip provided by GitHub directly.
```

> We encourage all Solidity developers to upgrade to version 0.8.18 to take advantage of these improvements and optimizations. Our team remains dedicated to continuously enhancing the Solidity platform to benefit the entire Ethereum ecosystem.

Reviewer: Are we a platform now? :)

Author: Old habits... 🤦

> If you want to perform a source build, please only use [`solidity_0.8.18.tar.gz`](https://github.com/ethereum/solidity/releases/download/v0.8.18/solidity_0.8.18.tar.gz) and not the zip provided by GitHub directly.

Reviewer: I'd switch to this wording:

```suggestion
If you want to perform a source build, please only use [`solidity_0.8.18.tar.gz`](https://github.com/ethereum/solidity/releases/download/v0.8.18/solidity_0.8.18.tar.gz) and not the source archives generated automatically by GitHub.
```

I already did that for the release notes.

> We encourage all Solidity developers to upgrade to version 0.8.18 to take advantage of these improvements and optimizations. Our team remains dedicated to continuously enhancing Solidity to benefit the entire Ethereum ecosystem.

Reviewer 2: ```suggestion
We recommend all Solidity developers to upgrade to version 0.8.18 to take advantage of these improvements and optimizations. Our team remains dedicated to continuously enhancing Solidity to benefit the entire Ethereum ecosystem.
```
Because we encourage all solidity users to...` was just used in the last sentence.

> We encourage all Solidity developers to upgrade to version 0.8.18 to take advantage of these improvements and optimizations. Our team remains dedicated to continuously enhancing Solidity to benefit the entire Ethereum ecosystem.

Reviewer: This bit might need some paragraph breaks.

Reviewer: What about this one? It's a wall of text currently :)

> We encourage all Solidity developers to upgrade to version 0.8.18 to take advantage of these improvements and optimizations. Our team remains dedicated to continuously enhancing the Solidity platform to benefit the entire Ethereum ecosystem.

Author: ```suggestion
We encourage all Solidity developers to upgrade to version 0.8.18 to take advantage of these improvements and optimizations. Our team remains dedicated to continuously enhancing Solidity to benefit the entire Ethereum ecosystem.
```

*Note*: `block.difficulty` is planned to be completely removed in Solidity version `0.9.0`.

## Full Changelog

#### Important fixes
*
*

### Bugfixes
* Docs build crash with sphinx-a4doc >= 1.4.0 #13820
* Required boost version in the docs does not match the code after recent changes #13870
* AST source locations for named mapping keys. #13864
* 

We especially thank all the contributors that made this release possible:
LIST

To upgrade to the latest version of the Solidity Compiler, simply follow the [installation instructions](https://docs.soliditylang.org/en/v0.8.18/installing-solidity.html) available in our documentation. 
Our team has made sure to provide detailed and straightforward steps to make the upgrade process as seamless as possible. If you have any questions or run into any issues during the upgrade process, don't hesitate to reach out to our [community Matrix channel](https://matrix.to/#/#ethereum_solidity:gitter.im). 
You can download the new version of Solidity here: [v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18).
If you want to perform a source build, please only use `solidity_0.8.18.tar.gz` and not the zip provided by GitHub directly.

> 

Author: ```suggestion
Please note that for those using Solidity versions below 0.8.0, [there are breaking changes](https://docs.soliditylang.org/en/v0.8.17/080-breaking-changes.html?highlight=update). We highly recommend reviewing the detailed list of breaking changes in our documentation to ensure a smooth upgrade process. Additionally, we encourage all Solidity users to regularly check for updates to stay up-to-date with the latest improvements and optimizations.
```

Bugfixes:
 * Parser: Disallow several ``indexed`` attributes for the same event parameter.
 * Parser: Disallow usage of the ``indexed`` attribute for modifier parameters.
 * SMTChecker: Fix display error for negative integers that are one more than powers of two.
 * SMTChecker: Fix internal error on chain assignments using static fully specified state variables.
 * SMTChecker: Fix internal error on multiple wrong SMTChecker natspec entries.
 * SMTChecker: Fix internal error when a public library function is called internally.
 * SMTChecker: Fix internal error when deleting struct member of function type.
 * SMTChecker: Fix internal error when using user-defined types as mapping indices or struct members.
 * SMTChecker: Improved readability for large integers that are powers of two or almost powers of two in error messages.
 * TypeChecker: Fix bug where private library functions could be attached with ``using for`` outside of their declaration scope.
 * Yul Optimizer: Hash hex and decimal literals according to their value instead of their representation, improving the detection of equivalent functions.

> 

Author: ```suggestion

*Note*: `block.difficulty` is planned to be completely removed in Solidity version `0.9.0`.

```

---
layout: post
published: true
title: 'Solidity 0.8.18 Release Announcement'
date: '2023-02-01'
author: Solidity Team
category: Releases
---

Introducing the newest version of the Solidity Compiler!
We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18). 
This latest version includes a range of improvements and it also introduces support for the [Paris upgrade](https://blog.ethereum.org/2022/08/24/mainnet-merge-announcement)!

> *Note*: `block.difficulty` is planned to be completely removed in Solidity version `0.9.0`.

Author: ```suggestion
```

---
layout: post
published: true
title: 'Solidity 0.8.18 Release Announcement'
date: '2023-02-01'
author: Solidity Team
category: Releases
---

Introducing the newest version of the Solidity Compiler!
We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18). 
This latest version includes a range of improvements and it also introduces support for the [Paris upgrade](https://blog.ethereum.org/2022/08/24/mainnet-merge-announcement)!

> 

Author: ```suggestion

### Compiler Features

* A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now, it was only possible to exclude the hash of the JSON metadata file (`--metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.
* The support to the Paris hard fork includes: 
    * Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
    * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.

*Note*: `block.difficulty` is planned to be completely removed in Solidity version `0.9.0`.
```

## Full Changelog

> Language Features:

Author: ```suggestion
### Language Features:
```

> ### Language Features:

Author: ```suggestion
### Language Features
```

 * Allow named parameters in mapping types.

> Compiler Features:

Author: ```suggestion
### Compiler Features
```

 * Optimizer: Added optimization rule ``and(shl(X, Y), shl(X, Z)) => shl(X, and(Y, Z))``.
 * Parser: More detailed error messages about invalid version pragmas.
 * SMTChecker: Make ``z3`` the default solver for the BMC and CHC engines instead of all solvers.
 * SMTChecker: Support Eldarica as a Horn solver for the CHC engine when using the CLI option ``--model-checker-solvers eld``. The binary ``eld`` must be available in the system.
 * Solidity Upgrade Tool: Remove ``solidity-upgrade`` tool.
 * Standard JSON: Add a boolean field ``settings.metadata.appendCBOR`` that skips CBOR metadata from getting appended at the end of the bytecode.
 * TypeChecker: Warn when using deprecated builtin ``selfdestruct``.
 * Yul EVM Code Transform: Generate more optimal code for user-defined functions that always terminate a transaction. No return labels will be pushed for calls to functions that always terminate.
 * Yul Optimizer: Allow replacing the previously hard-coded cleanup sequence by specifying custom steps after a colon delimiter (``:``) in the sequence string.
 * Yul Optimizer: Eliminate ``keccak256`` calls if the value was already calculated by a previous call and can be reused.

> Bugfixes:

Author: ```suggestion
### Bugfixes
```

We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18). 
This latest version includes a range of improvements and it also introduces support for the [Paris upgrade](https://blog.ethereum.org/2022/08/24/mainnet-merge-announcement)!

### Compiler Features

* A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now, it was only possible to exclude the hash of the JSON metadata file (`--metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.
* The support to the Paris hard fork includes: 
    * Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).

>     * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.

Author: ```suggestion
    * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.
    * `selfdestruct` is now considered deprecated ([EIP-6049](https://eips.ethereum.org/EIPS/eip-6049)) and the compiler will warn about its use, both in Solidity and in Yul, including inline assembly. There is currently no replacement, but its use is highly discouraged because it will eventually change its semantics and all contracts using it will be affected in some way.
```

 * Yul Optimizer: Hash hex and decimal literals according to their value instead of their representation, improving the detection of equivalent functions.

As always, a big thank you to all contributors who helped make this release possible!

## How to install/upgrade
To upgrade to the latest version of the Solidity Compiler, simply follow the [installation instructions](https://docs.soliditylang.org/en/v0.8.18/installing-solidity.html) available in our documentation. 
Our team has made sure to provide detailed and straightforward steps to make the upgrade process as seamless as possible. If you have any questions or run into any issues during the upgrade process, don't hesitate to reach out to our [community Matrix channel](https://matrix.to/#/#ethereum_solidity:gitter.im). 
You can download the new version of Solidity here: [v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18).
If you want to perform a source build, please only use [`solidity_0.8.18.tar.gz`](https://github.com/ethereum/solidity/releases/download/v0.8.18/solidity_0.8.18.tar.gz) and not the source archives generated automatically by GitHub.

> Please note that for those using Solidity versions below 0.8.0, [there are breaking changes](https://docs.soliditylang.org/en/v0.8.18/080-breaking-changes.html). We highly recommend reviewing the detailed list of breaking changes in our documentation to ensure a smooth upgrade process. Additionally, we encourage all Solidity users to regularly check for updates to stay up-to-date with the latest improvements and optimizations.

Author: ```suggestion

Please note that for those using Solidity versions below 0.8.0, [there are breaking changes](https://docs.soliditylang.org/en/v0.8.18/080-breaking-changes.html). We highly recommend reviewing the detailed list of breaking changes in our documentation to ensure a smooth upgrade process. Additionally, we encourage all Solidity users to regularly check for updates to stay up-to-date with the latest improvements and optimizations.
```

---
layout: post
published: true
title: 'Solidity 0.8.18 Release Announcement'
date: '2023-02-01'
author: Solidity Team
category: Releases
---

## Introducing the newest version of the Solidity Compiler!
We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18). 
This latest version includes a range of improvements and it also introduces support for the [Paris upgrade](https://blog.ethereum.org/2022/08/24/mainnet-merge-announcement)!

### Notable New Features

> * A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now, it was only possible to exclude the hash of the JSON metadata file (`--metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.

Author: ```suggestion
#### --no-cbor-metadataCommand-Line Flag
 A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now, it was only possible to exclude the hash of the JSON metadata file (`--metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.
```

* The support to the Paris hard fork includes: 
    * Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
    * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.

> * `selfdestruct` is now considered deprecated ([EIP-6049](https://eips.ethereum.org/EIPS/eip-6049)) and the compiler will warn about its use, both in Solidity and in Yul, including inline assembly. There is currently no replacement, but its use is highly discouraged because it will eventually change its semantics and all contracts using it will be affected in some way.

Author: ```suggestion
* `selfdestruct` is now considered deprecated ([EIP-6049](https://eips.ethereum.org/EIPS/eip-6049)) and the compiler will warn about its use, both in Solidity and in Yul, including inline assembly. There is currently no replacement, but its use is highly discouraged because it will eventually change its semantics and all contracts using it will be affected in some way.
* This release also involves a change in the generation of Yul code to EVM bytecode when enabling the optimizer: for Yul functions that never return back to the caller (since they always revert or always terminate the EVM call), we no longer push a return label to the stack like it is needed for other functions to jump back to the caller. This will decrease both gas cost and code size. Tooling that relies on analyzing our calling conventions on the bytecode level may need to be adjusted to accommodate this change.
```

>     * `selfdestruct` is now considered deprecated ([EIP-6049](https://eips.ethereum.org/EIPS/eip-6049)) and the compiler will warn about its use, both in Solidity and in Yul, including inline assembly. There is currently no replacement, but its use is highly discouraged because it will eventually change its semantics and all contracts using it will be affected in some way.

Author: ```suggestion 
* `selfdestruct` is now considered deprecated ([EIP-6049](https://eips.ethereum.org/EIPS/eip-6049)) and the compiler will warn about its use, both in Solidity and in Yul, including inline assembly. There is currently no replacement, but its use is highly discouraged because it will eventually change its semantics and all contracts using it will be affected in some way.
```

 * SMTChecker: Improved readability for large integers that are powers of two or almost powers of two in error messages.
 * TypeChecker: Fix bug where private library functions could be attached with ``using for`` outside of their declaration scope.
 * Yul Optimizer: Hash hex and decimal literals according to their value instead of their representation, improving the detection of equivalent functions.

As always, a big thank you to all contributors who helped make this release possible!

## How to install/upgrade
To upgrade to the latest version of the Solidity Compiler, simply follow the [installation instructions](https://docs.soliditylang.org/en/v0.8.18/installing-solidity.html) available in our documentation. 
Our team has made sure to provide detailed and straightforward steps to make the upgrade process as seamless as possible. If you have any questions or run into any issues during the upgrade process, don't hesitate to reach out to our [community Matrix channel](https://matrix.to/#/#ethereum_solidity:gitter.im). 
You can download the new version of Solidity here: [v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18).

> If you want to perform a source build, please only use [`solidity_0.8.18.tar.gz`](https://github.com/ethereum/solidity/releases/download/v0.8.18/solidity_0.8.18.tar.gz) and not the source archives generated automatically by GitHub.

Author: ```suggestion
If you want to build from the source code, do not use the source archives generated automatically by GitHub, instead please use [`solidity_0.8.18.tar.gz`](https://github.com/ethereum/solidity/releases/download/v0.8.18/solidity_0.8.18.tar.gz) and take a look at [our documentation on how to build from source](https://docs.soliditylang.org/en/v0.8.18/installing-solidity.html#building-from-source) if you need more help.
```
What about this? @cameel and @r0qs

Reviewer: Sounds good to me.

> If you want to perform a source build, please only use [`solidity_0.8.18.tar.gz`](https://github.com/ethereum/solidity/releases/download/v0.8.18/solidity_0.8.18.tar.gz) and not the source archives generated automatically by GitHub.

Reviewer 3: Not sure why we recommend "only use..." here. They can build from the source using the repository as well. Maybe rephrase to something that also include the repository or point to the building from source section in the docs (https://docs.soliditylang.org/en/v0.8.18/installing-solidity.html#building-from-source)?

Author: @r0qs actually has a valid point in terms of explainability. Let me try something.

---
layout: post
published: true
title: 'Solidity 0.8.18 Release Announcement'
date: '2023-02-01'
author: Solidity Team
category: Releases
---

Introducing the newest version of the Solidity Compiler!
We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18). 
This latest version includes a range of improvements and it also introduces support for the [Paris upgrade](https://blog.ethereum.org/2022/08/24/mainnet-merge-announcement)!

> ### Compiler Features

Author: ```suggestion
### Noteworthy Features
```

#### --no-cbor-metadataCommand-Line Flag
 A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now, it was only possible to exclude the hash of the JSON metadata file (`--metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.

> * The support for the Paris hard fork, which includes: 
>     * Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
>     * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.
>     *Note*: `block.difficulty` is planned to be completely removed in Solidity version `0.9.0`.

Author: ```suggestion

#### Support for Paris Hardfork
The support for the Paris hard fork, includes: 
    * Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
    * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.
    *Note*: `block.difficulty` is planned to be removed entirely in Solidity version `0.9.0`.
```

* This release also involves a change in the generation of EVM bytecode from Yul code when the optimizer is enabled: for Yul functions that never return back to the caller (i.e. always revert or always terminate the EVM call), the compiler no longer pushes a return label. The label is only used to jump back to the caller. This will decrease both gas cost and code size. Tooling that relies on the current calling convention on the bytecode level may need to be adjusted to accommodate this change.

## Full Changelog
### Language Features
 * Allow named parameters in mapping types.

> 

Author: ```suggestion
```

---
layout: post
published: true
title: 'Solidity 0.8.18 Release Announcement'
date: '2023-02-01'
author: Solidity Team
category: Releases
---

> Introducing the newest version of the Solidity Compiler!

Author: ```suggestion
## Introducing the newest version of the Solidity Compiler!
```

We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18). 
This latest version includes a range of improvements and it also introduces support for the [Paris upgrade](https://blog.ethereum.org/2022/08/24/mainnet-merge-announcement)!

> ### Noteworthy Features

Author: @cameel take a look

> ### Noteworthy Features

Reviewer 4: ```suggestion
### Notable New Features
```

In previews release announcements, we have referred to this as "Notable New Features" and then had subtitles for each of the Notable New Features. See here for reference: https://blog.soliditylang.org/2022/06/15/solidity-0.8.15-release-announcement/

If we want to re-name and re-structure how that section was usually done that's fine with me. Just pointing out that we should stick to one format moving forward and I'm personally not sure, if the bullet point list is a good tool for going into things more detailed.

Reviewer 4: If we wanted to structure it more closely to how we did in the past, I'd call the subsections:

`--no-cbor-metadata`Command-Line Flag
Support for Paris Hardfork
Deprecation of SELFDESTRUCT

Reviewer: Yes, I agree that sticking to the previous format and having subsections would be nice. Definitely better than the bullet points.

* A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now, it was only possible to exclude the hash of the JSON metadata file (`--metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.

> * The support to the Paris hard fork includes: 

Reviewer: ```suggestion
* The support for the Paris hard fork, which includes: 
```

    * Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
    * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.
* `selfdestruct` is now considered deprecated ([EIP-6049](https://eips.ethereum.org/EIPS/eip-6049)) and the compiler will warn about its use, both in Solidity and in Yul, including inline assembly. There is currently no replacement, but its use is highly discouraged because it will eventually change its semantics and all contracts using it will be affected in some way.

> * This release also involves a change in the generation of Yul code to EVM bytecode when enabling the optimizer: for Yul functions that never return back to the caller (since they always revert or always terminate the EVM call), we no longer push a return label to the stack like it is needed for other functions to jump back to the caller. This will decrease both gas cost and code size. Tooling that relies on analyzing our calling conventions on the bytecode level may need to be adjusted to accommodate this change.

Reviewer: I'd tweak the wording a bit:
```suggestion
* This release also involves a change in the generation of EVM bytecode from Yul code when the optimizer is enabled: for Yul functions that never return back to the caller (i.e. always revert or always terminate the EVM call), the compiler no longer pushes a return label. The label is only used to jump back to the caller. This will decrease both gas cost and code size. Tooling that relies on the current calling convention on the bytecode level may need to be adjusted to accommodate this change.
```

> *Note*: `block.difficulty` is planned to be completely removed in Solidity version `0.9.0`.

Reviewer: Shouldn't this be another bullet point under Paris changes?

Author: Deprecation deserve a distinguished position.

Author: In fact, I wonder if we have a warning box or something similar to really make this evident.

Reviewer: I think in markdown we don't.

Author: 🤔 then we can improve the discoverability. WDYT?

`*Note*: As part of the support for the Paris hard fork we are planning to completely remove`block.difficulty` in Solidity version `0.9.0`.`

Reviewer: But now it looks like a quote :)

I think it will be fine as if was if we just put it in a section together with the other things about Paris. It only looks weird because everything else is on a list. With sections it won't be a problem.

Bugfixes:
 * Parser: Disallow several ``indexed`` attributes for the same event parameter.
 * Parser: Disallow usage of the ``indexed`` attribute for modifier parameters.
 * SMTChecker: Fix display error for negative integers that are one more than powers of two.
 * SMTChecker: Fix internal error on chain assignments using static fully specified state variables.
 * SMTChecker: Fix internal error on multiple wrong SMTChecker natspec entries.
 * SMTChecker: Fix internal error when a public library function is called internally.
 * SMTChecker: Fix internal error when deleting struct member of function type.
 * SMTChecker: Fix internal error when using user-defined types as mapping indices or struct members.
 * SMTChecker: Improved readability for large integers that are powers of two or almost powers of two in error messages.
 * TypeChecker: Fix bug where private library functions could be attached with ``using for`` outside of their declaration scope.
 * Yul Optimizer: Hash hex and decimal literals according to their value instead of their representation, improving the detection of equivalent functions.

> #### Important fixes
> *
> *
> 
> ### Bugfixes
> * Docs build crash with sphinx-a4doc >= 1.4.0 #13820
> * Required boost version in the docs does not match the code after recent changes #13870
> * AST source locations for named mapping keys. #13864
> * 
> 
> 
> We especially thank all the contributors that made this release possible:
> LIST

Author: ```suggestion
We especially thank all the contributors that made this release possible:
Alexander Arlt, Alex Beregszaszi, andy53, Anton Paymyshev, Bhargava Shastry, Big-Aaron, Bojidar00, Bulgantamir Gankhuyag, chriseth, Christian Parpart, ChrisXXXXXXX, Damian Wechman, Daniel Kirchner, Doggo, Duc Thanh Nguyen, Franco Victorio, Franziska Heintel, George Plotnikov, hrkrshnn, Ikko Ashimine, Ishtiaque Zahid, John Kane, Kaan Uzdoğan, Kamil Śliwak, Leo Alt, ligi, Lokesh Kumar, Matheus Aguiar, Mathias L. Baumann, Mike Leach, Miles Liu, Minebuu, Mio, Nathaniel Jensen, Nikola Matić, Nishant Sachdeva, Nuno Santos, omahs, Paweł Bylica, Phill, Pierre Grimaud, Prusakova Katya, Rafal Stozek, Rajkumar gaur, Rhythm Bansal, Riley, Rodrigo Q. Saramago, Sabnock, Saw-mon-and-Natalie, Sebastian Supreme, Soham Zemse, Vinay, vlad, William Entriken, Yusuf Benli
```

> ### Acknowledgements 

Reviewer 4: In the past, we did not include the long version of acknowledgements here. To keep it short, we said:
> A big thank you to all contributors who helped make this release possible!

My personal preference is to keep it short, but if we are in favor of changing it to the long list, we should do so as an "active" decision and then stick to it from now on.

Author: I always include this in the release blogposts. It's a nicer way of thanking the community and getting them enough recognition.

Author: However, no strong opinion here. Happy to remove it.

> ### Acknowledgements 
> We especially thank all the contributors that made this release possible:
> Alexander Arlt, Alex Beregszaszi, andy53, Anton Paymyshev, Bhargava Shastry, Big-Aaron, Bojidar00, Bulgantamir Gankhuyag, chriseth, Christian Parpart, ChrisXXXXXXX, Damian Wechman, Daniel Kirchner, Doggo, Duc Thanh Nguyen, Franco Victorio, Franziska Heintel, George Plotnikov, hrkrshnn, Ikko Ashimine, Ishtiaque Zahid, John Kane, Kaan Uzdoğan, Kamil Śliwak, Leo Alt, ligi, Lokesh Kumar, Matheus Aguiar, Mathias L. Baumann, Mike Leach, Miles Liu, Minebuu, Mio, Nathaniel Jensen, Nikola Matić, Nishant Sachdeva, Nuno Santos, omahs, Paweł Bylica, Phill, Pierre Grimaud, Prusakova Katya, Rafal Stozek, Rajkumar gaur, Rhythm Bansal, Riley, Rodrigo Q. Saramago, Sabnock, Saw-mon-and-Natalie, Sebastian Supreme, Soham Zemse, Vinay, vlad, William Entriken, Yusuf Benli

Author: ```suggestion
As always, a big thank you to all contributors who helped make this release possible!
```

> Alexander Arlt, Alex Beregszaszi, andy53, Anton Paymyshev, Bhargava Shastry, Big-Aaron, Bojidar00, Bulgantamir Gankhuyag, chriseth, Christian Parpart, ChrisXXXXXXX, Damian Wechman, Daniel Kirchner, Doggo, Duc Thanh Nguyen, Franco Victorio, Franziska Heintel, George Plotnikov, hrkrshnn, Ikko Ashimine, Ishtiaque Zahid, John Kane, Kaan Uzdoğan, Kamil Śliwak, Leo Alt, ligi, Lokesh Kumar, Matheus Aguiar, Mathias L. Baumann, Mike Leach, Miles Liu, Minebuu, Mio, Nathaniel Jensen, Nikola Matić, Nishant Sachdeva, Nuno Santos, omahs, Paweł Bylica, Phill, Pierre Grimaud, Prusakova Katya, Rafal Stozek, Rajkumar gaur, Rhythm Bansal, Riley, Rodrigo Q. Saramago, Sabnock, Saw-mon-and-Natalie, Sebastian Supreme, Soham Zemse, Vinay, vlad, William Entriken, Yusuf Benli

Reviewer: By the way, for whatever reason we were including these only in release notes in github. I don't mind changing that if you think that's better but just FYI.

## How to install/upgrade
To upgrade to the latest version of the Solidity Compiler, simply follow the [installation instructions](https://docs.soliditylang.org/en/v0.8.18/installing-solidity.html) available in our documentation. 
Our team has made sure to provide detailed and straightforward steps to make the upgrade process as seamless as possible. If you have any questions or run into any issues during the upgrade process, don't hesitate to reach out to our [community Matrix channel](https://matrix.to/#/#ethereum_solidity:gitter.im). 
You can download the new version of Solidity here: [v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18).
If you want to perform a source build, please only use [`solidity_0.8.18.tar.gz`](https://github.com/ethereum/solidity/releases/download/v0.8.18/solidity_0.8.18.tar.gz) and not the source archives generated automatically by GitHub.

> Please note that for those using Solidity versions below 0.8.0, [there are breaking changes](https://docs.soliditylang.org/en/v0.8.17/080-breaking-changes.html?highlight=update). We highly recommend reviewing the detailed list of breaking changes in our documentation to ensure a smooth upgrade process. Additionally, we encourage all Solidity users to regularly check for updates to stay up-to-date with the latest improvements and optimizations.

Reviewer 4: This section is somewhat misleading, since the last breaking release (0.8.0) is a loong time ago. I see the idea behind this section though.

Reviewer: Yeah, got a bit surprised by it too when rereading now. But I think it would be fine, as long as it's separated from this big block of text. I'd just say `0.7.x and earlier` rather than `below 0.8.0` for clarity. Looked like typo to me at first.

> Please note that for those using Solidity versions below 0.8.0, [there are breaking changes](https://docs.soliditylang.org/en/v0.8.17/080-breaking-changes.html?highlight=update). We highly recommend reviewing the detailed list of breaking changes in our documentation to ensure a smooth upgrade process. Additionally, we encourage all Solidity users to regularly check for updates to stay up-to-date with the latest improvements and optimizations.

Reviewer: ```suggestion
Please note that for those using Solidity versions below 0.8.0, [there are breaking changes](https://docs.soliditylang.org/en/v0.8.18/080-breaking-changes.html). We highly recommend reviewing the detailed list of breaking changes in our documentation to ensure a smooth upgrade process. Additionally, we encourage all Solidity users to regularly check for updates to stay up-to-date with the latest improvements and optimizations.
```

We recommend all Solidity developers to upgrade to version 0.8.18 to take advantage of these improvements and optimizations. Our team remains dedicated to continuously enhancing Solidity to benefit the entire Ethereum ecosystem.

> Get ready to unleash your creativity and build even more impactful decentralized applications with the latest version of the Solidity Compiler. Happy coding!

Reviewer 4: ```suggestion
Get the new version of Solidity [here](https://github.com/ethereum/solidity/releases/tag/v0.8.18).
```
I would personally get rid of this sentence, and maybe close with the link to Github instead. That's also a style preference/ question of course,  but from my perspective Release Announcements can be a bit more clean/informational/professional compared to other blog posts, which can have different/friendlier tones.

If we decide to change the structure and tones of the release announcements that's okay, but again, we should stick to one style then moving forward.

Author: It's the idea behind being me doing them.

### Notable New Features

#### --no-cbor-metadataCommand-Line Flag
 A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now, it was only possible to exclude the hash of the JSON metadata file (`--metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.

#### Support for Paris Hardfork
The support for the Paris hard fork, includes: 
    * Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
    * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.
    *Note*: `block.difficulty` is planned to be removed entirely in Solidity version `0.9.0`.

> * `selfdestruct` is now considered deprecated ([EIP-6049](https://eips.ethereum.org/EIPS/eip-6049)) and the compiler will warn about its use, both in Solidity and in Yul, including inline assembly. There is currently no replacement, but its use is highly discouraged because it will eventually change its semantics and all contracts using it will be affected in some way.

Author: ```suggestion
#### Deprecation of `selfdestruct`
`selfdestruct` is now considered deprecated ([EIP-6049](https://eips.ethereum.org/EIPS/eip-6049)) and the compiler will warn about its use, both in Solidity and in Yul, including inline assembly. There is currently no replacement, but its use is highly discouraged because it will eventually change its semantics and all contracts using it will be affected in some way.

```

#### Support for Paris Hardfork
The support for the Paris hard fork, includes: 
    * Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
    * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.
    *Note*: `block.difficulty` is planned to be removed entirely in Solidity version `0.9.0`.
#### Deprecation of `selfdestruct`
`selfdestruct` is now considered deprecated ([EIP-6049](https://eips.ethereum.org/EIPS/eip-6049)) and the compiler will warn about its use, both in Solidity and in Yul, including inline assembly. There is currently no replacement, but its use is highly discouraged because it will eventually change its semantics and all contracts using it will be affected in some way.

> * This release also involves a change in the generation of EVM bytecode from Yul code when the optimizer is enabled: for Yul functions that never return back to the caller (i.e. always revert or always terminate the EVM call), the compiler no longer pushes a return label. The label is only used to jump back to the caller. This will decrease both gas cost and code size. Tooling that relies on the current calling convention on the bytecode level may need to be adjusted to accommodate this change.

Author: ```suggestion
#### EVM bytecode generation from Yul 
This release also involves a change in the generation of EVM bytecode from Yul code when the optimizer is enabled: for Yul functions that never return back to the caller (i.e. always revert or always terminate the EVM call), the compiler no longer pushes a return label. 
The label is only used to jump back to the caller. This will decrease both gas cost and code size. 
Tooling that relies on the current calling convention on the bytecode level may need to be adjusted to accommodate this change.
```

### Notable New Features

#### --no-cbor-metadataCommand-Line Flag
 A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now, it was only possible to exclude the hash of the JSON metadata file (`--metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.

#### Support for Paris Hardfork
The support for the Paris hard fork, includes: 
    * Deprecation of global `block.difficulty` and removal of `difficulty()` in inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
    * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.
    *Note*: `block.difficulty` is planned to be removed entirely in Solidity version `0.9.0`.
#### Deprecation of `selfdestruct`

> `selfdestruct` is now considered deprecated ([EIP-6049](https://eips.ethereum.org/EIPS/eip-6049)) and the compiler will warn about its use, both in Solidity and in Yul, including inline assembly. There is currently no replacement, but its use is highly discouraged because it will eventually change its semantics and all contracts using it will be affected in some way.

Author: ```suggestion
`selfdestruct` is now considered deprecated ([EIP-6049](https://eips.ethereum.org/EIPS/eip-6049)) and the compiler will warn about its use, both in Solidity and in Yul, including inline assembly. 
There is currently no replacement, but its use is highly discouraged because it will eventually change its semantics and all contracts using it will be affected in some way.
```

---
layout: post
published: true
title: 'Solidity 0.8.18 Release Announcement'
date: '2023-02-01'
author: Solidity Team
category: Releases
---

## Introducing the newest version of the Solidity Compiler!
We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18). 
This latest version includes a range of improvements and it also introduces support for the [Paris upgrade](https://blog.ethereum.org/2022/08/24/mainnet-merge-announcement)!

### Notable New Features

> #### --no-cbor-metadataCommand-Line Flag
>  A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. Until now, it was only possible to exclude the hash of the JSON metadata file (`--metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.

Author: ```suggestion
#### --no-cbor-metadata Command-Line Flag
 A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. 
 Until now, it was only possible to exclude the hash of the JSON metadata file (`--metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.
```

    * Introduction of global `block.prevrandao` in Solidity and `prevrandao()` in inline assembly for EVM versions >= Paris.
    *Note*: `block.difficulty` is planned to be removed entirely in Solidity version `0.9.0`.
#### Deprecation of `selfdestruct`
`selfdestruct` is now considered deprecated ([EIP-6049](https://eips.ethereum.org/EIPS/eip-6049)) and the compiler will warn about its use, both in Solidity and in Yul, including inline assembly. There is currently no replacement, but its use is highly discouraged because it will eventually change its semantics and all contracts using it will be affected in some way.

#### EVM bytecode generation from Yul 
This release also involves a change in the generation of EVM bytecode from Yul code when the optimizer is enabled: for Yul functions that never return back to the caller (i.e. always revert or always terminate the EVM call), the compiler no longer pushes a return label. 
The label is only used to jump back to the caller. This will decrease both gas cost and code size. 
Tooling that relies on the current calling convention on the bytecode level may need to be adjusted to accommodate this change.

> 

Author: ```suggestion
```

 * SMTChecker: Fix display error for negative integers that are one more than powers of two.
 * SMTChecker: Fix internal error on chain assignments using static fully specified state variables.
 * SMTChecker: Fix internal error on multiple wrong SMTChecker natspec entries.
 * SMTChecker: Fix internal error when a public library function is called internally.
 * SMTChecker: Fix internal error when deleting struct member of function type.
 * SMTChecker: Fix internal error when using user-defined types as mapping indices or struct members.
 * SMTChecker: Improved readability for large integers that are powers of two or almost powers of two in error messages.
 * TypeChecker: Fix bug where private library functions could be attached with ``using for`` outside of their declaration scope.
 * Yul Optimizer: Hash hex and decimal literals according to their value instead of their representation, improving the detection of equivalent functions.

#### Important fixes
*
*

### Bugfixes
* Docs build crash with sphinx-a4doc >= 1.4.0 #13820
* Required boost version in the docs does not match the code after recent changes #13870
* AST source locations for named mapping keys. #13864

> * 

Reviewer 3: This seciton is duplicated.

We especially thank all the contributors that made this release possible:

> LIST

Reviewer 3: Now we can get the list from here: https://github.com/ethereum/solidity/releases/tag/v0.8.18

---
layout: post
published: true
title: 'Solidity 0.8.18 Release Announcement'
date: '2023-02-01'
author: Solidity Team
category: Releases
---

## Introducing the newest version of the Solidity Compiler!
We are excited to announce the latest release of the Solidity Compiler, [Solidity v0.8.18](https://github.com/ethereum/solidity/releases/tag/v0.8.18). 
This latest version includes a range of improvements and it also introduces support for the [Paris upgrade](https://blog.ethereum.org/2022/08/24/mainnet-merge-announcement)!

### Notable New Features

> #### --no-cbor-metadata Command-Line Flag

Reviewer: ```suggestion
#### `--no-cbor-metadata` Command-Line Flag
```
Looks weird in preview otherwise.

Reviewer: Alternatively:
```suggestion
#### Disabling CBOR metadata
```

Author: You were faster. As I was editing it. 😛

 A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. 
 Until now, it was only possible to exclude the hash of the JSON metadata file (`--metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.

#### Support for Paris Hardfork
The support for the Paris hard fork, includes: 
* Deprecation of global `block.difficulty` built-in in Solidity and removal of `difficulty()` instruction from inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
* Introduction of global `block.prevrandao` built-in in Solidity and `prevrandao()` instruction in inline assembly for EVM versions >= Paris.

>     *Note*: `block.difficulty` is planned to be removed entirely in Solidity version `0.9.0`.

Author: ```suggestion

*Note*: `block.difficulty` is planned to be removed entirely in Solidity version `0.9.0`.
```

### Notable New Features

#### Disabling CBOR metadata
 A new command-line flag (`--no-cbor-metadata`) and Standard JSON option (`settings.metadata.appendCBOR: false`) to prevent compiler from appending the CBOR metadata section at the end of the bytecode. 
 Until now, it was only possible to exclude the hash of the JSON metadata file (`--metadata-hash none`/`settings.metadata.bytecodehash: "none"`) from that section, but the bytecode would still include a small bit of CBOR data containing the version of the compiler.

#### Support for Paris Hardfork
The support for the Paris hard fork, includes: 
* Deprecation of global `block.difficulty` built-in in Solidity and removal of `difficulty()` instruction from inline assembly for EVM versions >= Paris. The change is due to the renaming and semantic changes of the `difficulty` instruction introduced by [EIP-4399](https://eips.ethereum.org/EIPS/eip-4399).
* Introduction of global `block.prevrandao` built-in in Solidity and `prevrandao()` instruction in inline assembly for EVM versions >= Paris.

*Note*: `block.difficulty` is planned to be removed entirely in Solidity version `0.9.0`.

> #### Deprecation of `selfdestruct`

Author: ```suggestion

#### Deprecation of `selfdestruct`
```
